wip: disable/enable doc endpoints

jordigh opened the pull request:
## Context

wip

## Proposed solution

add the endpoints

## Related issues

Kinda like adding `removedAt`

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help <!-- Detail how we can help you -->

## Screenshots / Screencasts

<!-- delete if not relevant -->

paulfitz commented on app/gen-server/lib/homedb/Interfaces.ts:
```diff
@@ -104,6 +104,7 @@ export interface DocAuthResult {
   access: roles.Role|null;    // The access level for the requesting user. Null on error.
   removed: boolean|null;      // Set if the doc is soft-deleted. Users may still have access
                               // to removed documents for some purposes. Null on error.
+  disabled: boolean|null;     // Like removed, but users do not have access to it. Null on error.
```
That's a good explanation of the fundamental difference I think...

jordigh replied:
Hm,isn't that exactly what I did? Added those two endpoints?

paulfitz commented on app/gen-server/migration/1759434763338-DocDisabledAt.ts:
```diff
@@ -0,0 +1,20 @@
+        type: nativeValues.dateTimeType,
+        isNullable: true,
+      }));
+      await queryRunner.createIndex("docs", new TableIndex({
```
What is the purpose of this index?

jordigh replied:
We have several queries where `disabled_at = NULL` in SQL, so the index seems necessary. The corresponding `removed_at` column also has an index and the queries for the two properties are very similar.

jordigh commented on app/common/UserAPI.ts:
```diff
@@ -423,6 +424,8 @@ export interface UserAPI {
   deleteDoc(docId: string): Promise<void>;      // delete doc permanently
   softDeleteDoc(docId: string): Promise<void>;  // soft-delete doc
   undeleteDoc(docId: string): Promise<void>;    // recover soft-deleted doc
+  disableDoc(docId: string): Promise<void>;     // soft-delete doc but hide it more and forbid it more
```
Okay, reworded

jordigh commented on app/gen-server/lib/homedb/HomeDBManager.ts:
Thanks.

jordigh commented on app/gen-server/lib/homedb/HomeDBManager.ts:
```diff
@@ -843,7 +843,7 @@ export class HomeDBManager {
       .innerJoin('workspaces.org', 'orgs')
       .where('docs.workspace_id = workspaces.id')
-      .andWhere('workspaces.removed_at IS NULL AND docs.removed_at IS NULL');
+      .andWhere('workspaces.removed_at IS NULL AND docs.removed_at IS NULL AND docs.disabled_at IS NULL');
```
How's that comment?

jordigh commented on app/gen-server/lib/homedb/HomeDBManager.ts:
```diff
@@ -1095,10 +1096,15 @@ export class HomeDBManager {
-        (doc.removedAt === null && doc.workspace.removedAt === null) :
-        (doc.removedAt || doc.workspace.removedAt)) {
+    const docRemoved = Boolean(doc.removedAt || doc.workspace.removedAt);
+    const onlyShowRemoved = Boolean(scope.showRemoved && !scope.showAll);
```
I like your version a little bit better but it still has a ternary inside an `if` condition, which still looks like a translation of an `XOR`.

jordigh commented on app/gen-server/lib/homedb/HomeDBManager.ts:
```diff
@@ -5117,22 +5135,32 @@ export class HomeDBManager {
       }
-      const doc: Document = this.unwrapQueryResult(await verifyEntity(docQuery));
-      if (!removedAt) {
+      // For `disabledAt`, the API endpoint already checks for admin access
```
I've decided it's cleaner to grant at the start of the API call a `specialPermit` to the document in question since we just passed an admin check.

jordigh commented on app/gen-server/lib/homedb/HomeDBManager.ts:
```diff
@@ -979,6 +979,7 @@ export class HomeDBManager {
         .leftJoinAndSelect('billing_account.product', 'product')
         .where('key = :key', {key: shareKey})
         .andWhere('doc.removed_at IS NULL')
+        .andWhere('doc.disabled_at IS NULL')
```
I'm sorry for not understanding, but since the permissions of a fork are based on the original document, if the original is is gone, shouldn't the fork also be gone?

Also, what do you mean by propagating the flag to the caller? Which flag?

jordigh commented:
Tests are failing right now because Github Actions is building a bad merge of `main` and `jordigh/disable-doc`, which is something I just learned it does when you a trigger workflow on the `pull_request` event. A rebase should fix that, but I don't want to rebase in the middle of a code review.

berhalak commented on app/gen-server/lib/homedb/HomeDBManager.ts:
```diff
@@ -3760,7 +3774,8 @@ export class HomeDBManager {
         .leftJoinAndSelect('doc_acl_rules.group', 'doc_groups')
         .leftJoinAndSelect('doc_groups.memberUsers', 'doc_users')
-        .andWhere('docs.removed_at IS NULL')  // Don't grant guest access for soft-deleted docs.
+        .andWhere('docs.removed_at IS NULL')   // Don't grant guest access for soft-deleted docs.
```
Hmm, I think that the fact that there are some docs in the workspace doesn't change anything in permissions we have. Permissions are not calculated on the fly, they are persisted in the doc. If there are no documents in the workspace, the permission should be removed together with the last document (for example move endpoint is updating those permissions).

paulfitz commented on app/server/lib/DocApi.ts:
```diff
@@ -1169,6 +1172,19 @@ export class DocWorkerApi {
 
+    // POST /api/docs/:docId/disable
+    // Disables doc (similar to soft-delete, but needs admin access)
+    this._app.post('/api/docs/:docId/disable', requireInstallAdmin, async (req, res) => {
```
Why are `/enable` and `/disable` implemented in different locations? Aren't they similar?

paulfitz commented on test/gen-server/lib/disabledAt.ts:
```diff
@@ -0,0 +1,247 @@
+      doc31 = await api.newDoc({name: 'doc31'}, ws3);
+    });
+
+    it('hides soft-deleted docs from regular api', async function() {
```
soft-deleted? is that  accurate?

jordigh commented:
I've reworked this PR.

Disabled documents now only disallow access and modification, but allow being deleted and they show up in the document list, where they can be deleted or undeleted.

<img width="1366" height="200" alt="image" src="https://github.com/user-attachments/assets/59c14e4d-bfc3-4b64-9881-16ae89f0c1c9" />

 Disabled documents are not garbage collected, unless they are also deleted.

I rewrote the tests to reflect these changes.

 Reworked commit messages:
 
 * d04ba029fec6 Document: add new disabledAt column to the entity

  This necessitates adding the property to many related interfaces.

* 0465d0c90be6 AuditEvents: add document.enable/document.disable events

  We will be adding endpoints that enable and disable documents, and we need to be able to log this fact.

* ed2e8fe43e43 HomeDBManager: disable access to disabled docs

  When asserting access, this now checks if the document is disabled,
  and if so, denies access, unless the document is being removed.

* c1b6eb1b0758 HomeDBManager: add functions to disable/enable documents

  These functions only set or clear the `disabledAt` property.

* aebc32f2d0f5 HomeDBManager: disallow moving or modifying disabled docs

  If a doc is disabled, it should not be modifiable. Only deletable.

* 6c5f8b3b7284 UserAPI: add enable/disable methods

  This is mostly just for tests, for feature parity with similar UserAPI
  methods.

* 1ef7cbe76b7b DocApi: add `POST /api/docs/:docId/disable` and enable endpoints

  Note that these endpoints use the same `requireInstallAdmin`
  middleware and do no API throttling, mostly because the throttling
  logic is a bit complicated for a document that is being disabled.
  Denying all non-admin requests throttles sufficiently.

* 6695cc5e5486 pruneAPIResult: don't prune `disabledAt`

  On second thought (re: c49a6124866ba00bb5c70c3ac6c90dc7d8044045) I do
  want to have `disabledAt` showing through the API. I'm going to use it
  in the UI to indicate if a document is disabled or not.

* fe763e31046d DocMenu: also show if a document was disabled

  In order of priority, we show if a document was removed, disabled, or
  just active and updated. This logic is now centralised in
  `getUpdatedAt`.

* a70feec8ee82 DocMenu: also sort docs by `disabledAt` time

  As before, the priority is removed, then disabled, then regular active
  docs.

* f17e654f1f13 DocList: show a lock icon on disabled docs

* b5add25435bb DocList: disallow URLs for disabled docs

  Users will get a 403 anyway if they try to follow the doc URL, but no
  need to cruelly tricking them into thinking they could get there from
  a disabled doc.

* 6ed4eb79935f tests: add test to verify that disabledAt works in browser, API and Housekeeper

berhalak commented on test/nbrowser/disabledAt.ts:
```diff
@@ -0,0 +1,149 @@
+  it('lets admin disable a document via API', async function() {
+    await adminApi.disableDoc(docId);
+    assert.typeOf((await adminApi.getDoc(docId)).disabledAt, 'string');
+    await gu.waitForServer();
```
This wait is not needed here.

jordigh commented on test/nbrowser/disabledAt.ts:
```diff
@@ -0,0 +1,149 @@
+    await enabledDoc.findWait('.test-dm-doc-options', 500).click();
+    assert.isFalse(await gu.findOpenMenuItem('li', /Move/).matches('[class*=disabled]'));
+    assert.isFalse(await gu.findOpenMenuItem('li', /Rename/).matches('[class*=disabled]'));
+    await gu.sendKeys(Key.ESCAPE);
```
Ah, thanks, I've added these calls.

jordigh commented on app/server/lib/DocApi.ts:
```diff
@@ -2394,6 +2429,33 @@ export class DocWorkerApi {
+      details: {
+        document: {
+          ..._.pick(document, "id", "name"),
+          workspace: _.pick(document.workspace, "id", "name"),
```
There was no good reason. I've just added the workspace to both. I was copying the removed/restore logging, but there it made a bit more sense removing a document "moves" it from a workspace to the trash and restoring it "moves" it from the trash to a workspace (or so it appears to the user).

Since disabling can't reasonably be called moving in any way, I've just decided to add the workspace to both disabling and enabling.

jordigh replied:
Okay, done that.

berhalak commented on test/nbrowser/disabledAt.ts:
```diff
@@ -0,0 +1,148 @@
+  });
+
+  after(async function() {
+    oldEnv.restore();
```
You should restart server here, I think server in nbrowser tests is shared.

paulfitz commented on app/server/lib/Authorizer.ts:
```diff
@@ -600,6 +600,11 @@ export function assertAccess(
     throw new ErrorWithCode("AUTH_NO_VIEW", "Document is deleted", {status: 404});
   }
 
+  // Disabled docs in the process of being removed (deleted) are allowed,
```
What is being allowed, specifically?

jordigh replied:
Deleting or undeleting a doc. I've rephrased.

paulfitz commented on app/server/lib/DocApi.ts:
```diff
@@ -2320,6 +2325,41 @@ export class DocWorkerApi {
+      },
+      disable: {
+        dbMethod: this._dbManager.disableDocument.bind(this._dbManager),
+        logEvent: this._logDisableDocumentEvents.bind(this),
```
The log methods look like they could be combined, as do the enable/disableDocument methods, and if they were combined then this method could be simpler?

jordigh commented on app/server/lib/DocApi.ts:
```diff
@@ -2320,6 +2325,41 @@ export class DocWorkerApi {
+    const mreq = req as RequestWithLogin;
+    const docId = req.params.docId || req.params.did;
+
+    // We have admin access, so grant a special permit to this doc
```
Okay, I added a docstring.

jordigh commented on test/nbrowser/disabledAt.ts:
```diff
@@ -0,0 +1,149 @@
+    const enabledDoc = entries[0];
+    const disabledDoc = entries[1];
+
+    await delay(0);
```
Stray debug code, removed it.

jordigh commented on app/server/lib/requestUtils.ts:
```diff
@@ -28,7 +28,7 @@ export const TEST_HTTPS_OFFSET = process.env.GRIST_TEST_HTTPS_OFFSET ?
   'apiKey', 'billingAccountId', 'firstLoginAt', 'lastConnectionAt', 'filteredOut', 'ownerId', 'gracePeriodStart',
   'stripeCustomerId', 'stripeSubscriptionId', 'stripeProductId', 'userId', 'isFirstTimeUser', 'allowGoogleLogin',
-  'authSubject', 'usage', 'createdBy', 'unsubscribeKey', 'disabledAt',
+  'authSubject', 'usage', 'createdBy', 'unsubscribeKey'
```
I've edited the corresponding [commit message](https://github.com/gristlabs/grist-core/pull/1865/commits/802c8d9c043bf269fcdd1b8d75204e194771faed).
